include/BareCpper/ATsamd/ExternalInterruptController_ATsamd5x.hpp

```diff
@@ -0,0 +1,182 @@
+          // Wait for clock synchronization
+				  while (GCLK->SYNCBUSY.reg);
+          // Reset the EIC
+          reset();
+          // Enable the EIC
+          enable();
+        }
+
+        using CallbackT = std::function<void()>;
+        
+        template<typename PinT>
+        void attachInterrupt(const PinT& pin
+                            , const CallbackT& callback
+                            , const ExternalInterruptSense& mode
+                            , const ExternalInterruptDebouncer& debounce = ExternalInterruptDebouncer::Off)
+        {
+          // registers are enable-protected, disable EIC before configuring if it is enabled
+          const bool shouldRestart = EIC->CTRLA.reg & EIC_CTRLA_ENABLE;
+          if(shouldRestart) disable();
+          // set the Pin alternative function to peripheral A (EIC)
+          BareCpper::gpioFunction<PinT>(pin, std::optional<ATsamd5x::Peripheral>{ATsamd5x::Peripheral::A});
+          // configure the EIC sense mode
+          constexpr auto channel = (PinT::id().pin > 15) ? (PinT::id().pin - 16) : PinT::id().pin;
+          constexpr uint8_t configRegNum = (channel > 7);
+          constexpr uint8_t configRegShift = (channel - 8*configRegNum) << 2; //< channel config starts at bit 4^channel
+          EIC->CONFIG[configRegNum].reg &= ~(0x7 << configRegShift);
+          EIC->CONFIG[configRegNum].reg |= (static_cast<uint8_t>(mode) << configRegShift);
+          // configure debouncer
+          if(debounce == ExternalInterruptDebouncer::Off)
+          {
+            EIC->DEBOUNCEN.reg &= ~(0x1 << channel);
+          }
+          else
+          {
+            constexpr uint8_t dpresclareRegShift = configRegNum << 2;  //< dpreslacer for pins 0-7 start at bit 0, for pins 8-15 at bit 4 (4^n)
+            // set debounce prescaler
+            EIC->DPRESCALER.reg &= ~(0x7 << dpresclareRegShift);
+            EIC->DPRESCALER.reg |= (static_cast<uint8_t>(debounce) << dpresclareRegShift);
+            EIC->DEBOUNCEN.reg |= (0x1 << channel);
+          }
+          // set callback
+          ExternalInterruptController::callbacks_[channel] = callback;
+          // enable interrupt in NVIC
+          constexpr auto irqNumber = getIrqNumber(channel);
+          NVIC_ClearPendingIRQ(irqNumber);
+          NVIC_SetPriority(irqNumber, 0);
+          NVIC_EnableIRQ(irqNumber);
+          // enable interrupt in EIC
+          EIC->INTENSET.reg = (0x1 << channel);
+          // reenable EIC
+          if(shouldRestart) enable();
+        }
+
+        template<typename PinT>
+        void detachInterrupt(const PinT& pin)
+        {
+          // registers are enable-protected, disable EIC before configuring if it is enabled
+          const bool shouldRestart = EIC->CTRLA.reg & EIC_CTRLA_ENABLE;
+          if(shouldRestart) disable();
+          constexpr auto channel = (PinT::id().pin > 15) ? (PinT::id().pin - 16) : PinT::id().pin;
```

Comment:
Comment: Would be useful to document 'port'  is not used as the EXINT mux design limits the use of the same index pin on different ports i.e. "SAMD mux for EIC is shared across each port e.g. PA01, PB01, PC01 cannot be configured at the same time independently. Also the lower 16 are overlayed with the higher 16 pins of each port e.g. PA01 and PA16 cannot be configured at the same time"
  ... not sure this wording is quite clear, could do some work.

Reply:
I've added doc comments for the functions now. We can make some kind of manager that will allocate/free resources, and it will be useful for many things (event system, dmac, timers, etc.)